src/qp_problems.py

```diff
@@ -0,0 +1,193 @@
+import numpy as np
+import pandas as pd
+import qpsolvers
+from qpsolvers import solve_qp
+import scipy
+from helper_functions import nearestPD
+from covariance import Covariance
+from constraints import Constraints
+from optimization_data import OptimizationData
+
+
+
+# This class converts an financial optimization problem to a standard quadratic optimization.
+# This is the last step before passing problems to solvers.
+
+
+class QuadraticProgram(dict):
+
+    def __init__(self, *args, **kwargs):
+        super(QuadraticProgram, self).__init__(*args, **kwargs)
+        self['sparse'] = self.get('sparse') is None
+
+    def linearize_turnover_constraint(self,
+                                      x_init: np.ndarray,
+                                      to_budget = float('inf')) -> None:
+        # Dimensions
+        n = len(self.get('q'))
+        m = 0 if self.get('G') is None else self.get('G').shape[0]
+
+        # Objective
+        if self.get('P') is not None:
+            P = np.zeros(shape = (2*n, 2*n))
+            P[0:n, 0:n] = self.get('P')
+            P = nearestPD(P)
+        q = np.append(self.get('q'), np.zeros(n))
+
+        # Inequality constraints
+        G = np.zeros(shape = (m+2*n+1, 2*n))
+        if self.get('G') is not None:
+            G[0:m, 0:n] = self.get('G')
+        G[m:(m+n), 0:n] = np.eye(n)
+        G[m:(m+n), n:(2*n)] = np.eye(n) * (-1)
+        G[(m+n):(m+2*n), 0:n] = np.eye(n) * (-1)
+        G[(m+n):(m+2*n), n:(2*n)] = np.eye(n) * (-1)
+        G[(m+2*n), ] = np.append(np.zeros(n), np.ones(n))
+        h = self.get('h') if self.get('h') is not None else np.empty(shape=(0,))
+        h = np.append(h, np.append(np.append(x_init, x_init * (-1)), to_budget))
+
+        # Equality constraints
+        if self.get('A').ndim == 1:
+            A = np.append(self.get('A'), np.zeros(n))
+        else:
+            A = np.zeros(shape = (self.get('A').shape[0], 2*n))
+            A[0:self.get('A').shape[0], 0:n] = self.get('A')
+
+        # Override the original matrices
+        self['P'] = P
```

Comment:
This is common in all 3 functions, could be a helper function that all functions call it?